### PR TITLE
Build process

### PR DESCRIPTION
**Pull request by mediocretes**
Update the Instrumental.NET package to make it easier to work with:
- A completely functional script/setup
- A script/build that builds the correct packages and warns you if steps were missed (eventually, I'd like to have it do the steps for you, but one does what one can)
- A directory layout where all the directories have just one purpose

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,2 +1,5 @@
 ### 0.0.1 [April ??, 2016]
 * Initial release
+
+### 0.1.0 [April 13, 2016]
+* Clean up existing build process
```
Comment (jqr) [inline CHANGELOG.md:2]: Can we put new items on top in here?

File: .gitignore [hunk 2]
```diff
@@ -172,3 +172,6 @@ UpgradeLog*.htm
 
 # Microsoft Fakes
 FakesAssemblies/
+
+# vendor dir for packages from nuget
+vendor/*
```
Comment (jqr) [inline .gitignore]: If we make this `vendor/*/*` it should achieve the same result but not mask some of the files we already have checked in under vendor.

File: script/build [hunk 3]
```diff
@@ -12,14 +12,19 @@ which mcs > /dev/null  || (echo "mcs must be installed"; exit 1)
 
 script/nuget install vendor/packages.config -o vendor
 mkdir -p bin
-
-references=""
+pwd
```
Comment (jqr) [inline script/build]: Debug line?

File: CHANGELOG.md (hunk 1 above)
Comment (jqr) [inline CHANGELOG.md:2]: Looks like the version is actually 0.1.1 as well.

**Comment by jqr** [comment]
This seems fine, but can you provide a description for this pull request?

**Comment by jqr** [comment]
Can you explain a bit what you mean by "building the correct packages" and "missing steps" as this build script was nearly identical to the [docraptor-csharp one](https://github.com/DocRaptor/docraptor-csharp/blob/master/script/build). Do we need to make changes there too?

**Comment by mediocretes** [comment]
It was definitely not building any packages due to missing references and other issues, perhaps "correct" should have been omitted.

My comment about missing steps refers entirely to the versioning - there are three authoritative sources of versioning, and automating keeping them in sync seems like the wrong move right now, so I'm at least reporting on what they are on every package build.  See also c5d450261c48277365db40f2399db0031b4729b3 where "build" is broken into "compile" and "package" and nuget install is part of script setup (as well as the package step).

**Comment by mediocretes** [comment]
That being said, script/setup + script/build on docraptor definitely does not work.  I will fix it or make an issue right now.

**Comment by mediocretes** [comment]
Also, due to bin being excluded, this is definitely still not the right thing WRT vendor and dependencies... I'll fix that

File: script/build [hunk 4]
```diff
@@ -10,16 +10,21 @@ fi
 which mono > /dev/null || (echo "mono must be installed"; exit 1)
 which mcs > /dev/null  || (echo "mcs must be installed"; exit 1)
 
-script/nuget install vendor/packages.config -o vendor
+bin/nuget install packages.config -o vendor
```
Comment (jqr) [inline script/build]: `bin/nuget` isn't checked in.

Comment (mediocretes) [inline script/build]: Yeah, that's another oversight.  I really wanted executables in bin, but vendor turns out to be much more reasonable.  I'll fix it right up.

**Comment by janxious** [comment]
not sure what the state of this is, but it doesn't seem to produce an assembly that is usable in the csharp repl, so maybe some of the latter work superceded and replaces this PR?

**Comment by mediocretes** [comment]
Yes, repl usability is kinda fucked until later.  running script/test should compile the agent, and then compile the tests, and then run the tests.

I'll take your comments on later branches as a :+1: here :)

File: script/check_version [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+set -e
+
+echo
+echo "Built packages have the following versions:"
+echo $(cat CHANGELOG.md | sed -n 's/.*\(\([[:digit:]]\.\)\{2\}[[:digit:]]\).*/\1/p' | head -n 1
+) -- CHANGELOG.md
+echo $(monodis --assembly bin/Instrumental.dll | grep Version | sed -n 's/.*\(\([[:digit:]]\.\)\{3\}[[:digit:]]\).*/\1/p') -- Instrumental.dll '(from src/Properties/AssemblyInfo.cs)'
+echo $(ls bin | sed -n 's/Instrumental\.\(\([[:digit:]]\.\)\{2\}[[:digit:]]\)\.nupkg/\1/p' | tail -n 1) -- Nuget Package version '(Instrumental.nuspec)'
+echo "If these do not match, please correct the source and rebuild"
```
Comment (jqr) [inline script/check_version:10]: How close do these need to match? Mine currently reports:

```
0.2.0 -- CHANGELOG.md
0.2.0  -- Agent constant(from src/Agent.cs)
0.2.0.0 -- Instrumental.dll (from src/Properties/AssemblyInfo.cs)
0.2.0 -- Nuget Package version (Instrumental.nuspec)
```

Comment (mediocretes) [inline script/check_version:10]: They are not allowed to have any numbers to disagree.

That is to say, 0.2.0 and 0.2.0.0 are the same number, for these purposes.  What you are seeing is correct.  When the rate of change settles down, we should release 0.3.0.

Comment (jqr) [inline script/check_version:10]: Will the DLL will have 4 parts and the last one will always be `0`? At least in our use?

Comment (mediocretes) [inline script/check_version:10]: It should, yes.  It should be easy to truncate that, I don't think we will ever include build number.

**Pull request merged**